File: mytest.c
#include "mytest.h"

#include <stdio.h>
#include <string.h>

int print_string(const char *s) {
	printf("'%s'\n", s);
	return strlen(s);
}

void print_string_fix(const char *s, int l) {
	printf("%d characters in '%s'\n", l, s);
}

void get_name(char *s, int l) {
	const char name[] = "Guido";
	strncpy(s, name, l);
}

float area(const struct Rect *r) {
    return (r->a.x - r->b.x) * (r->a.y - r->b.y);
}

int rect_from_center(struct Rect *r, const struct Point *c, const float w, const float h) {
    r->a.x = c->x - w / 2;
    r->b.x = c->x + w / 2;
    r->a.y = c->y - h / 2;
    r->b.y = c->y + h / 2;
	return 42;
}

int multiply(const int w, const int h, float *result, const float *a, const float *b) {
	for (int i = 0; i < w * h; ++i) result[i] = a[i] * b[i];
	return 0;
}

_returns_(okay) int mysincos(double value, _out_ double *s, _out_ double *c) {
	*s = sin(value);
	*c = cos(value);
	return okay;
}

File: mytest.py
import numpy
import safectypes

lib = safectypes.load_dll("mytest.so", "mytest.h")

p1 = lib.Point(1.0, 2.0)
p2 = lib.Point(4.0, 1.0)
r1 = lib.Rect(p1, p2)
f1 = lib.area(r1)
print f1, (r1.a.x - r1.b.x) * (r1.a.y - r1.b.y);

p3 = lib.Point(5, 6)
w = 3
h = w
r2 = lib.rect_from_center(p3, h=h)
print (r2.a.x, r2.a.y, r2.b.x, r2.b.y), (p3.x - 0.5 * w, p3.y - 0.5 * h, p3.x + 0.5 * w, p3.y + 0.5 * h)

a = 2 * numpy.ones((3, 4), dtype=numpy.float32)
b = 3 * numpy.ones((3, 4), dtype=numpy.float32)
c = numpy.zeros((3, 4), dtype=numpy.float32)
lib.multiply(result=c, a=a, b=b)
print c

d = 4 * numpy.ones((3, 4), dtype=numpy.float32)
e = lib.multiply(a, d)
print e

print lib.mysincos(3.14)

